File: DSA/Assignment_4/insertion.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node
{
    int data;
    struct Node *next;
}Node;

int main()
{
    int n;
    printf("Enter number of nodes:");
    scanf("%d",&n);
    Node *head=NULL;
    Node *tail=NULL;
    for(int i=0;i<n;i++)
    {
        Node *cur=(Node *)malloc(sizeof(Node));
        printf("Enter element:");
        scanf("%d",&cur->data);
        cur->next=NULL;

        if(head==NULL)
        {
            head=cur;
            tail=cur;
        }
        else
        {
            tail->next=cur;
            tail=cur;
        }
    }

    printf("Original Linked List:");
    Node *temp=head;
    while(temp)
    {
        printf("%d ",temp->data);
        temp=temp->next;
    }
    printf("\n");

    int pos,ele,p=0;
    printf("Enter position and element to insert:");
    scanf("%d %d",&pos,&ele);
    Node *new_node=(Node *)malloc(sizeof(Node));
    new_node->data=ele;
    new_node->next=NULL;
    temp=head;
    while(temp && p!=pos)
    {
        temp=temp->next;
        p++;
    }
    new_node->next=temp->next;
    temp->next=new_node;

    printf("Modified Linked List:");
    temp=head;
    while(temp)
    {
        printf("%d ",&temp->data);
        temp=temp->next;
    }
    printf("\n");

    //Free memory
    temp=head;
    while(temp)
    {
        Node *next_node=temp->next;
        free(temp);
        
    }
    
    return 0;
}

File: DSA/Assignment_4/insertion.py
class Node:
    def __init__(self, data):
        self.data=data
        self.next=None

class Linked_list:
    def __init__(self, head):
        self.head=None

    def insert_ele(self, data):
        new_Node=Node(data)
        if self.head is None:
            self.head=new_Node
            return
        cur=self.head
        while cur.next:
            cur=cur.next
        cur.next=new_Node

    def print_list(self):
        temp=self.head
        while temp:
            print(temp.data,end=' ')
            temp=temp.next
        print()

    def insert_pos(self, ele, pos):
        new_node=Node(ele)
        temp=self.head
        p=1
        while temp:
            if pos==0:
                self.head=new_node
                self.head


list=Linked_list()

n=int(input("Enter number of elements:"))
for i in range(n):
    data=int(input("Enter an element:"))
    list.insert_ele(data)

print("Original List:",end=' ')
list.print_list()

ele=int(input("Enter element to insert:"))
pos=int(input("Enter position of element:"))
list.insert_pos(ele, pos)

print("Modified List:",end=' ')
list.print_list()